objects/src/assets/vault.rs

```diff
@@ -89,6 +92,43 @@ impl AssetVault {
     // PUBLIC MODIFIERS
     // --------------------------------------------------------------------------------------------
 
+    /// Applies the specified delta to the asset vault.
+    ///
+    /// # Errors
+    /// Returns an error:
+    /// - If the total value of assets is greater than or equal to 2^63.
+    /// - If the delta contains an addition/subtraction for a fungible asset that is not stored in
+    ///   the vault.
+    /// - If the delta contains a non-fungible asset removal that is not stored in the vault.
+    /// - If the delta contains a non-fungible asset addition that is already stored in the vault.
+    pub fn apply_delta(&mut self, delta: &AccountVaultDelta) -> Result<(), AssetVaultError> {
+        for (&faucet_id, &delta) in delta.fungible().iter() {
+            let asset = FungibleAsset::new(faucet_id, delta.unsigned_abs())
+                .expect("Not a fungible faucet ID or delta is too large");
+            if delta >= 0 {
+                self.add_fungible_asset(asset)?;
+            } else {
+                self.remove_fungible_asset(asset)?;
+            }
+        }
+
+        for (&asset, &action) in delta.non_fungible().iter() {
+            // SAFETY: the asset must be a valid word representation of a non-fungible asset.
+            //         This is safe because we allow only to add non-fungible assets.
+            let asset = unsafe { NonFungibleAsset::new_unchecked(asset.into()) };
```

Comment:
If we use `NonFungibleAsset` as the key in non-fungible asset delta, we should be able to avoid this transformation.